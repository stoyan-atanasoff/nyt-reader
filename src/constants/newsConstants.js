// Module News constants
export const NEWS_API_URL = 'https://api.nytimes.com/svc/mostpopular/v2/viewed/7.json?api-key=';
export const NEWS_API_KEY = 'tm1LRCluoKntuqAdrmR2gEP76MGI1aac';

export const FETCH_NEWS = 'NYT/NEWS/FETCH_NEWS';
export const FETCH_NEWS_SUCCESS = 'NYT/NEWS/FETCH_NEWS_SUCCESS';
export const FETCH_NEWS_FAIL = 'NYT/NEWS/FETCH_NEWS_FAIL';
