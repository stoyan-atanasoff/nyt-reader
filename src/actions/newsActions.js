import { FETCH_NEWS, FETCH_NEWS_SUCCESS, FETCH_NEWS_FAIL } from "../constants/newsConstants";

export const fetchNews = () => {
  return {
    type: FETCH_NEWS,
  }
};

export const fetchNewsSuccess = ( newsList ) => {
  return {
    type: FETCH_NEWS_SUCCESS,
    newsList
  }
};

export const fetchNewsFail = ( err ) => {
  return {
    type: FETCH_NEWS_FAIL,
    err,
  }
};
