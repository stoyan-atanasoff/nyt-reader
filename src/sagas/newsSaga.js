import { put, call, takeEvery } from 'redux-saga/effects';
import request from "../utils/request";
import {FETCH_NEWS, NEWS_API_KEY, NEWS_API_URL} from "../constants/newsConstants";
import {fetchNewsFail, fetchNewsSuccess} from "../actions/newsActions";

export function* fetchNews () {
  try {
    const URL = `${NEWS_API_URL}${NEWS_API_KEY}`;
    const query = yield call(request, URL);
    yield put(fetchNewsSuccess(query.results))
  } catch (e) {
    yield put(fetchNewsFail(e));
  }
}

export default function* newsSaga() {
  yield takeEvery(FETCH_NEWS, fetchNews);
}
