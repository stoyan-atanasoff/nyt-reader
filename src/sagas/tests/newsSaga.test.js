/**
 * Tests for HomePage sagas
 */

import { put, takeEvery } from 'redux-saga/effects';
import newsSaga, { fetchNews } from '../newsSaga';
import {fetchNewsFail, fetchNewsSuccess} from "../../actions/newsActions";
import {FETCH_NEWS} from "../../constants/newsConstants";

describe('fetch news saga', () => {
  let fetchNewsGenerator;

  beforeEach(() => {
    fetchNewsGenerator = fetchNews();

    const callDescriptor = fetchNewsGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should call the fetchNewsSuccess action ', () => {
    const putDescriptor = fetchNewsGenerator.next({ results: ['test successful result'] }).value;
    expect(putDescriptor).toEqual(put(fetchNewsSuccess(['test successful result'])));
  });

  it('should call the fetchNewsFail action if the request failed', () => {
    const response = new Error('Some error');
    const putDescriptor = fetchNewsGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(fetchNewsFail(response)));
  });

});

describe('news saga', () => {
  const saga = newsSaga();

  it('should start task to watch for FETCH_NEWS invokes', () => {
    const takeLatestDescriptor = saga.next().value;
    expect(takeLatestDescriptor).toEqual(takeEvery(FETCH_NEWS, fetchNews));
  });
});
