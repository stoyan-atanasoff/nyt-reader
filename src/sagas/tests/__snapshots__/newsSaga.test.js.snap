// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetch news saga should call the fetchNewsFail action if the request failed 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://api.nytimes.com/svc/mostpopular/v2/viewed/7.json?api-key=tm1LRCluoKntuqAdrmR2gEP76MGI1aac",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`fetch news saga should call the fetchNewsSuccess action  1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://api.nytimes.com/svc/mostpopular/v2/viewed/7.json?api-key=tm1LRCluoKntuqAdrmR2gEP76MGI1aac",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
